store/src/main/java/org/apache/rocketmq/store/kv/CompactionStore.java

```diff
@@ -0,0 +1,163 @@
+    private final ScheduledExecutorService compactionSchedule;
+    private final int compactionInterval;
+    private final int compactionThreadNum;
+    private final int offsetMapSize;
+
+    private static final InternalLogger log = InternalLoggerFactory.getLogger(LoggerName.STORE_LOGGER_NAME);
+
+    public CompactionStore(MessageStore defaultMessageStore) {
+        this.defaultMessageStore = defaultMessageStore;
+        this.compactionLogTable = new ConcurrentHashMap<>();
+        MessageStoreConfig config = defaultMessageStore.getMessageStoreConfig();
+        String storeRootPath = config.getStorePathRootDir();
+        this.compactionPath = Paths.get(storeRootPath, COMPACTION_DIR).toString();
+        this.compactionLogPath = Paths.get(compactionPath, COMPACTION_LOG_DIR).toString();
+        this.compactionCqPath = Paths.get(compactionPath, COMPACTION_CQ_DIR).toString();
+        this.positionMgr = new CompactionPositionMgr(compactionPath);
+        if (config.getCompactionThreadNum() <= 0) {
+            this.compactionThreadNum = Runtime.getRuntime().availableProcessors();
+        } else {
+            this.compactionThreadNum = config.getCompactionThreadNum();
+        }
+        this.compactionSchedule = Executors.newScheduledThreadPool(this.compactionThreadNum,
+            new ThreadFactoryImpl("compactionSchedule_"));
+        this.offsetMapSize = config.getMaxOffsetMapSize() / compactionThreadNum;
+
+        this.compactionInterval = defaultMessageStore.getMessageStoreConfig().getCompactionScheduleInternal();
+    }
+
+    public void load() {
+        File logRoot = new File(compactionLogPath);
+        File[] fileTopicList = logRoot.listFiles();
+        if (fileTopicList != null) {
+            for (File fileTopic : fileTopicList) {
+                if (!fileTopic.isDirectory()) {
+                    continue;
+                }
+
+                File[] fileQueueIdList = fileTopic.listFiles();
+                if (fileQueueIdList != null) {
+                    for (File fileQueueId : fileQueueIdList) {
+                        if (!fileQueueId.isDirectory()) {
+                            continue;
+                        }
+                        try {
+                            String topic = fileTopic.getName();
+                            int queueId = Integer.parseInt(fileQueueId.getName());
+
+                            if (Files.isDirectory(Paths.get(compactionCqPath, topic, String.valueOf(queueId)))) {
+                                CompactionLog log = new CompactionLog(defaultMessageStore, topic, queueId,
+                                    offsetMapSize, positionMgr, compactionLogPath, compactionCqPath);
+                                compactionLogTable.put(topic + "_" + queueId, log);
+                                compactionSchedule.scheduleWithFixedDelay(log::doCompaction, compactionInterval, compactionInterval, TimeUnit.SECONDS);
```

Comment:
Different compactionLog files are distinguished by topic and queueId, and do compaction in every compactionLog. Does this mean messages with the same keys must send to the same queue?

Reply:
Yes.